doc/release/upcoming_changes/26081.improvement.rst

```diff
@@ -0,0 +1,21 @@
+Shapes are now covariant
+------------------------
+When ``ndarray`` was made generic, more details and implementation were
+left to future development.  While it was recognized that shape was inherently
+a tuple (and thus covariant), development towards ``TypeVarTuple`` does not
+yet allow bounds or variance.  This change allows user code to type check in
+functions that attempt to specify the shape parameter:
```

Comment:
This comment about the type bounds is not quite accurate in the case of `ndarray`. The ndarray class was made generic prior to the development of `TypeVarTuple`, and as such what was effectively a placeholder slot was added for a future shape `TypeVar`. And while type var tuples do not support a bound, type vars *do*:

```diff
- _ShapeType = TypeVar("_ShapeType", covariant=True, bound=Any)
+ _ShapeType = TypeVar("_ShapeType", covariant=True, bound=tuple[int, ...])
```

Another option is to convert the first type var of `ndarray` into a type var tuple, but this has the very drawback that (to the extent of my knowledge) there's no way of representing a tuple with an unknown/arbitrary length (aka array dimensions).

Reply:
Thank you for the review!

> The `ndarray` class was made generic prior to the development of `TypeVarTuple`

Oh, sorry, I was just going off of git blame vs PEP date, and trying to fill in some context for readers who had heard of `TypeVarTuple` why this PR was moving away from that option:  changing `_ShapeType` to covariant pre-empts a PR changing it to a `TypeVarTuple` until the latter supports variance.  There's also more of my rationale at the end of this comment.  Would the following language work?

```
Static shape typing for `ndarray` is a long-term effort that continues 
with this change.  While shape is a variadic type variable, this change
chooses to keep `TypeVar` rather than `TypeVarTuple`, as the latter
does not yet allow bounds or variance.  This choice allows users
to type check functions that attempt to specify the shape parameter
as a `NewType`:
```

> And while type var tuples do not support a bound, type vars _do_

I'm 99% on board with bounding to `tuple[int, ...]`, as in your code and @jorenham's suggestion.  My only reservation is that I don't know enough about structured array data types.  E.g., in a length-3 structured array with two fields (or a record array of the same), would the shape typevar be `tuple[int, int]`?  Or is the dtype typevar responsible for the field length, meaning the shape typevar is `tuple[int]`?  Because I don't know, I limited this PR just towards variance, not bounds.  But if it's desired, I'm more than happy to add the bound.

> Another option is to convert the first type var of ``ndarray`` into a type var tuple, but this has the very drawback that (to the extent of my knowledge) there's no way of representing a tuple with an unknown/arbitrary length (aka array dimensions).

I thought `TypeVarTuple`s are always arbitrary length before they're narrowed/resolved?  But by gut feeling alone, `TypeVarTuple` seem less useful than `TypeVar("Shape", bound=tuple[int, ...])` until more of the TBDs from PEP 646 are resolved (e.g. [manipulating individual types][1] and [shape arithmetic][2]), especially because `Tuple` now also supports unpacking types.  I also prefer the behavior of `tuple` to `TypeVarTuple` when dealing with [unknown type parameters][3].  That is, I prefer type checking to show more shape errors, even if it requires peppering my code with `cast`.

[1]: https://peps.python.org/pep-0646/#overloads-for-accessing-individual-types
[2]: https://peps.python.org/pep-0646/#shape-arithmetic
[3]: https://peps.python.org/pep-0646/#unspecified-type-parameters-tuple-vs-typevartuple